### PR TITLE
fix vote and block insertion race condition

### PR DESCRIPTION
**Pull request by wjrjerome**

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 1]
```diff
@@ -604,10 +605,36 @@ func (x *XDPoS_v2) voteHandler(chain consensus.ChainReader, voteMsg *utils.Vote)
...
 */
-func (x *XDPoS_v2) onVotePoolThresholdReached(chain consensus.ChainReader, pooledVotes map[common.Hash]utils.PoolObj, currentVoteMsg utils.PoolObj) error {
+func (x *XDPoS_v2) onVotePoolThresholdReached(chain consensus.ChainReader, pooledVotes map[common.Hash]utils.PoolObj, currentVoteMsg utils.PoolObj, proposedBlockHeader *types.Header) error {
+
+	masternodes := x.GetMasternodes(chain, proposedBlockHeader)
+
+	// Filter out non-Master nodes signatures
+	var wg sync.WaitGroup
+	wg.Add(len(pooledVotes))
+	validSignatureMap := map[common.Hash]utils.Signature{}
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: question: is this `validSignatureMap` safe when many routines use it?

Comment (wgr523) [reply consensus/XDPoS/engines/engine_v2/engine.go]: if not, I think use `chan` is better

Comment (liam-lai) [reply consensus/XDPoS/engines/engine_v2/engine.go]: yes, it is safe if there is no risk on writing on same hash in same time. Also chan is hard to use on this case, chan is better if we know the number of item will put into the `chan`, but with current implementation, len of validSignatures is dynamic.

**Review by liam-lai — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 1 above)
Comment (wgr523) [reply consensus/XDPoS/engines/engine_v2/engine.go]: but I think it may write at same time, this line:
`validSignatureMap[hash] = v.Signature`

Comment (liam-lai) [reply consensus/XDPoS/engines/engine_v2/engine.go]: is this hash means block hash or vote hash?

Comment (liam-lai) [reply consensus/XDPoS/engines/engine_v2/engine.go]: if it's block hash, then we need to atomic this operation. If it's vote, then every vote from different nodes should have different hash

Comment (wgr523) [reply consensus/XDPoS/engines/engine_v2/engine.go]: ok, I googled about `map` and it says it's not safe...

**Review by wgr523 — APPROVED** [review]

**Pull request merged**